seo/builder.go

```diff
@@ -0,0 +1,375 @@
+package seo
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"reflect"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/qor5/admin/l10n"
+	h "github.com/theplant/htmlgo"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+var (
+	defaultGlobalSEOName = "Global SEO"
+)
+
+type (
+	contextVariablesFunc func(interface{}, *Setting, *http.Request) string
+	Option               func(*Builder)
+)
+
+func WithInherit(inherited bool) Option {
+	return func(b *Builder) {
+		b.inherited = inherited
+	}
+}
+
+func WithGlobalSEOName(name string) Option {
+	return func(b *Builder) {
+		name = GetSEOName(name)
+		if name == "" {
+			panic("The global seo name must be not empty")
+		}
+		b.seoRoot.name = name
+		delete(b.registeredSEO, defaultGlobalSEOName)
+		b.registeredSEO[name] = b.seoRoot
+	}
+}
+
+func WithLocales(locales ...string) Option {
+	return func(b *Builder) {
+		b.locales = locales
+	}
+}
+
+func NewBuilder(db *gorm.DB, ops ...Option) *Builder {
+	globalSEO := &SEO{name: defaultGlobalSEOName}
+	globalSEO.RegisterSettingVariables(struct {
+		SiteName string
+	}{})
+	globalSEO.RegisterPropFuncForOG(&PropFunc{
```

Comment:
globalSEO.RegisterOpenGraphProperty("og:url", 
                 func(_ interface{}, _ *Setting, req *http.Request) string {
			return req.URL.String()
		}
	)

Reply:
Decided to remove this default function and `PropFunc`. but keep `RegisterMetaProperty` function for explicit invoke